docs/implementation_design/event_driven_model.md

```diff
@@ -1 +1,72 @@
+
+* Event handler (callback) - is a function that is executed when an event of given type occurs. Event handler can accept multiple arguments (e.g. message ID or plugin) so that it can perform appropriate actions.
+  Every handler must accept the same set of parameters. Each event has different set of handlers (there can be multiple handlers) that are executed when event is triggered.
+
+* Trigger - is a method that triggers execution of event handlers with given parameter values.
+
+
+## Creating new event with custom callbacks
+
+Below are the steps that show the example code necessary to create a custom event, attach a handler and trigger the event. 
+
+1. Create function that will call event handlers (handler caller). This function is necessary in order to simplify calling handlers. 
+   Each event has only one handler, which means that all handlers must share the same interface.
+   Callers for all events must also share the same interface - the first argument represents handler function that will be called represented by generic argument.
+   Further arguments represent parameters that will be passed to handler during execution. Below are example callers that accept one and two parameters respectively. 
+   More arguments can be passed in similar manner. 
+   
+```go
+func sigleArgCaller(handler interface{}, params ...interface{}) {
+    handler.(func (*Plugin))(params[0].(*Plugin))
+}
+
+func twoArgsCaller(handler interface{}, params ...interface{}) {
+    handler.(func(*peer.Peer, error))(params[0].(*peer.Peer), params[1].(error))
+}
+```
+
+`handler.(func (*Plugin))(params[0].(*Plugin))` - this code seems a little complicated, so to make things simpler we will divide into smaller parts and explain each:
+
+* `handler.(func (*Plugin))` (A) - this part does type casting of handler onto type of desired function - in this case it's function that accepts `*Plugin` as its only parameters.
+* `params[0].(*Plugin)` (B)- similarly to previous part, first element of parameter slice is type-casted onto `*Plugin` type, so that it matches handler function interface.
+* `handler.(func (*Plugin))(params[0].(*Plugin))` - the whole expression calls the type-casted handler function with type-casted parameter value. We can also write this as `A(B)` to make things simpler.
+
+
+2. Next, a new event object needs to be created. We pass handler caller as an argument, which is saved inside the object to be called when event is triggered.
+
+```go
+import "github.com/iotaledger/hive.go/events"
+
+NewEvent := events.NewEvent(sigleArgCaller)
+```
+
+3. After creating the event, handlers (or callbacks) can be attached to it. An event can have multiple callbacks, however they all need to share the same interface. 
+   One thing to note, is that functions are not passed directly - first they are wrapped into a `events.Closure` object like in the example below. 
```

Comment:
What is the reason for adding such a wrapper?

Reply:
Apparently it does not work without it and the reason why has to do with the NewClosure creates a uintpr. Since the reason for this is external though (hive.go) we are probably ok to not need to elaborate here.